### PR TITLE
Review chap 9

### PR DESCRIPTION
**Pull request by maiq721**

### CODE REVIEW

File: app/assets/stylesheets/custom.scss [hunk 1]
```diff
@@ -163,3 +163,17 @@ input {
     color: $state-danger-text;
   }
 }
+
+.checkbox {
+  margin-top: -10px;
+  margin-bottom: 10px;
```
Comment (mdmahmudurrahman) [inline app/assets/stylesheets/custom.scss]: you can merge line 168 and 169 into 1 line

File: app/helpers/sessions_helper.rb [hunk 2]
```diff
@@ -1,10 +1,18 @@
...
     session[:user_id] = user.id
   end
 
   def current_user
-    @current_user ||= User.find_by id: session[:user_id]
+    if user_id = session[:user_id]
+      @current_user ||= User.find_by id: user_id
+    elsif user_id = cookies.signed[:user_id]
+      user = User.find_by id: user_ids
+      if user&.authenticated?(cookies[:remember_token])
```
Comment (mdmahmudurrahman) [inline app/helpers/sessions_helper.rb]: remove redundant ()

File: app/helpers/sessions_helper.rb [hunk 3]
```diff
@@ -15,4 +23,22 @@ def log_out
     session.delete :user_id
     @current_user = nil
   end
+
+  def remember user
+    user.remember
+    cookies.permanent.signed[:user_id] = user.id
```
Comment (mdmahmudurrahman) [inline app/helpers/sessions_helper.rb]: I have a question: What is the purpose of using signed cookie here? Why it is useful ?

File: app/helpers/sessions_helper.rb [hunk 4]
```diff
@@ -15,4 +23,22 @@ def log_out
...
+    cookies.permanent[:remember_token] = user.remember_token
+  end
+
+  def forget user
+    user.forget
+    cookies.delete(:user_id)
+    cookies.delete(:remember_token)
```
Comment (mdmahmudurrahman) [inline app/helpers/sessions_helper.rb]: Please remove redundant ()

File: app/helpers/sessions_helper.rb [hunk 5]
```diff
@@ -15,4 +23,22 @@ def log_out
...
+  end
+
+  def log_out
+    forget current_user
+    session.delete(:user_id)
```
Comment (mdmahmudurrahman) [inline app/helpers/sessions_helper.rb]: same here

File: app/models/user.rb [hunk 6]
```diff
@@ -1,10 +1,36 @@
...
   VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-]+(\.[a-z\d\-]+)*\.[a-z]+\z/i
   validates :email, presence: true, length: {maximum: Settings.validate.email.maximum},
     format: {with: VALID_EMAIL_REGEX},
     uniqueness: {case_sensitive: false}
   has_secure_password
   validates :password, presence: true, length: {minimum: Settings.validate.password.minimum}
+
+  def remember
+    self.remember_token = User.new_token
+    update_attribute :remember_digest, User.digest(remember_token)
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: 1. I have another question: Can you tell me all of the difference between **update_attribute** and **update_attributes** in active record
2. Use **update_attributes** here

File: app/models/user.rb [hunk 7]
```diff
@@ -1,10 +1,36 @@
...
+  end
+
+  class << self
+    def User.digest string
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: As you wrapped the class method inside the class << self block, no need to declare a class method using it class name. In summary remove **User** 
```ruby
def digest string
  # Method body
end 
```

File: app/models/user.rb [hunk 8]
```diff
@@ -1,10 +1,36 @@
...
+      cost = ActiveModel::SecurePassword.min_cost ? BCrypt::Engine::MIN_COST :
+        BCrypt::Engine.cost
+      BCrypt::Password.create string, cost: cost
+    end
+
+    def User.new_token
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: same here

File: app/helpers/sessions_helper.rb (hunk 3 above)
Comment (maiq721) [reply app/helpers/sessions_helper.rb]: use signed cookie to store session data(user id) can avoid an attack session hacking

File: app/models/user.rb (hunk 6 above)
Comment (maiq721) [reply app/models/user.rb]: The difference is: update_attribute bypasses the validations and update_attribute is used to update record with single attribute:
update_attribute(name, value)
 update_attributes is used to update record with multiple attributes
update_attributes(attributes)

File: app/models/user.rb (hunk 7 above)
Comment (maiq721) [reply app/models/user.rb]: I forget to remove User.. thanks

**Pull request merged**